cardano-node/src/Cardano/Node/Socket.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Node.Socket
+  ( gatherConfiguredSockets
+  , SocketConfigError(..)
+  , renderSocketConfigError
+  )
+where
+
+import qualified Prelude
+import           Prelude (String)
+import           Cardano.Prelude hiding (local)
+
+import           Control.Monad.Trans.Except.Extra (handleIOExceptT)
+import           System.Directory (removeFile)
+import           System.IO.Error (isDoesNotExistError)
+
+import           Cardano.Config.Types
+
+#ifdef SYSTEMD
+import           System.Systemd.Daemon (getActivatedSockets)
+#endif
+
+
+import           Network.Socket (Socket, AddrInfo (..), AddrInfoFlag (..),
+                    SocketType (..), defaultHints, getAddrInfo)
+
+
+
+-- | Errors for the current module.
+data SocketConfigError
+    = NoPublicSocketGiven
+    | NoLocalSocketGiven
+    | ClashingPublicSocketGiven
+    | ClashingLocalSocketGiven
+    | LocalSocketError FilePath IOException
+    | GetAddrInfoError NodeAddress IOException
+  deriving Show
+
+instance Exception SocketConfigError where
+  displayException = renderSocketConfigError
+
+renderSocketConfigError :: SocketConfigError -> String
+renderSocketConfigError NoPublicSocketGiven =
+    "No configuration for the node's public socket. Please specify a socket "
+ <> "path either in the config file, on the command line or via systemd socket "
+ <> "activation."
+
+renderSocketConfigError NoLocalSocketGiven =
+    "No configuration for the node's local socket. Please specify a socket "
+ <> "path either in the config file, on the command line or via systemd socket "
+ <> "activation."
+
+renderSocketConfigError ClashingPublicSocketGiven =
+    "Configuration for the node's public socket supplied both by config/cli and "
+ <> "via systemd socket activation. Please use one or the other but not both."
+
+renderSocketConfigError ClashingLocalSocketGiven =
+    "Configuration for the node's local socket supplied both by config/cli and "
+ <> "via systemd socket activation. Please use one or the other but not both."
+
+renderSocketConfigError (LocalSocketError fp ex) =
+    "Failure while attempting to remove the stale local socket: "
+ <> fp <> " : " <> displayException ex
+
+renderSocketConfigError (GetAddrInfoError addr ex) =
+    "Failure while getting address information for the public listening "
+ <> "address: " <> show addr <> " : " <> displayException ex
+
+
+-- | Gather from the various sources of configuration which sockets we will use
+-- for the public node-to-node and the local node-to-client IPC.
+--
+-- We get such configuration from:
+--
+-- * node config file
+-- * node cli
+-- * systemd socket activation
+--
+gatherConfiguredSockets :: NodeConfiguration
+                        -> NodeCLI
+                        -> ExceptT SocketConfigError IO
+                                   (Either [Socket] [AddrInfo],
```

Comment:
Returning a tuple of `Either`s as the result is confusing. It would be much better if you created a custom data type IMO.

Reply:
Agree with @Jimbo4350

Reply:
Done.